### ISSUE TITLE
Issue with requestIP when using body validator

### ISSUE THREAD

**Issue by eygsoft** [elysiajs/elysia#377]
**Description:**
Hello,

I have encountered an issue while using `requestIP` in conjunction with a body validator in my application. It seems that when a request uses a body validator, the `requestIP` method called within the `onTransform` callback returns `null`.

I suspect that this might be due to the request being closed after body validation, which results in `requestIP` returning `null`.

**Versions:**
- Bun Version: 1.0.18
- Elysia Version: 0.8.3

**Test Code:**
```typescript
import { Elysia, t } from 'elysia'
const app = new Elysia()
app
  // Always available
  .onRequest(({ request }) => {
    console.log(app.server!.requestIP(request)?.address)
  })
  // Not available when using body validator
  .onTransform(({ request }) => {
    console.log(app.server!.requestIP(request)?.address) // Not available when using body validator
  })
  // Similar to onTransform
  .derive(({ request }) => ({ ip: app.server!.requestIP(request)?.address }))
  // foo ::ffff:127.0.0.1
  .post('/foo/:id', ({ ip }) => `foo ${ip}`, {
    query: t.Object({ name: t.String() }), // Query and params do not affect the issue
    params: t.Object({ id: t.Numeric() })
  })
  // bar undefined
  .post('/bar', ({ ip }) => `bar ${ip}`, {
    body: t.Object({ num: t.Integer() })
  })
  .listen(3000)
```

**Additional Information:**
I also reviewed the source code of the `Elysia IP` and `Rate Limit` community plugins listed on the official website. Both of these plugins rely on `app.server!.requestIP(request)` for their functionality. As a result, when using a body validator, they both become ineffective and do not return the expected IP address.

This issue seems to impact multiple community plugins that rely on the `requestIP` method when used in conjunction with body validators.

**Comment by a-rebets** [comment]
I confirm this behaviour - [link to Codesandbox](https://codesandbox.io/p/devbox/elysia-ip-bug-ytwdy4?embed=1&file=%2Freadme.md)

The global hook used by the [elysia-rate-limit](https://github.com/rayriffy/elysia-rate-limit) _for the main logic_ is `onBeforeHandle`, while [elysia-ip](https://github.com/gaurishhs/elysia-ip) uses the `derive` method (hook?), which should be invoked during the Transform phase of the lifecycle

I included both these methods in the testing script and these are the logs of the server:
```text
-- new request (GET) --
<onRequest> ip: ::ffff:127.0.0.1
-- new request (POST) --
<onRequest> ip: ::ffff:127.0.0.1
<onBeforeHandle> ip: ::ffff:127.0.0.1
-- new request (POST) --
<onRequest> ip: ::ffff:127.0.0.1
<onBeforeHandle> ip: undefined
```
Simulated client:
```text
Called the first POST endpoint - 200, body - foo ::ffff:127.0.0.1
Called the second POST endpoint - 200, body - bar undefined
```

Honestly I think i understand "the Lifecycle" well enough, having spent countless hours in the docs, but here... I don't quite get it. The headers of the original don't change from `onRequest` to `onBeforeHandle`, but (obviously) the "bodyUsed" parameter changes since Elysia reads the body during the Validation phase:
<details>

<summary>parameters.json</summary>

```jsonc
[
    {
        // onRequest hook
        "bodyUsed": false,
        "headers": {
            "content-type": "application/json",
            "connection": "keep-alive",
            "user-agent": "Bun/1.0.23",
            "accept": "*/*",
            "host": "localhost:3000",
            "accept-encoding": "gzip, deflate, br",
            "content-length": "11"
        }
    },
    {
        // derive method (Transform event)
        "bodyUsed": true,
        "headers": {
            "content-type": "application/json",
            "connection": "keep-alive",
            "user-agent": "Bun/1.0.23",
            "accept": "*/*",
            "host": "localhost:3000",
            "accept-encoding": "gzip, deflate, br",
            "content-length": "11"
        }
    },
    {
        // onBeforeHandle hook
        "bodyUsed": true,
        "headers": {
            "content-type": "application/json",
            "connection": "keep-alive",
            "user-agent": "Bun/1.0.23",
            "accept": "*/*",
            "host": "localhost:3000",
            "accept-encoding": "gzip, deflate, br",
            "content-length": "11"
        }
    }
]
```

</details>

And yes, docs explicitly say that `onRequest` is the best hook to use for working with IP / analytics / etc. So I suppose the creators of `elysia-ip` / `elysia-rate-limit`, @rayriffy and @gaurishhs, have their reasoning behind the hook choice (side note - great plugins overall!), but I'm still attaching this quote from the docs:
<blockquote>
The first life-cycle event to get executed for every new request is recieved.

As onRequest is designed to provide only the most crucial context to reduce overhead, it is recommended to use in the following scenario:  
- Caching  
- Rate Limiter / IP/Region Lock  
- Analytic  
- Provide custom header, eg. CORS
</blockquote>

I could submit a PR to both these repos, if you guys think switching is a good idea.

### Real reason

I suspect it's just a Bun bug, cause I managed to reproduce oven-sh/bun#6613

It's due to the weird behaviour of Bun's API, when, if you wait for any Promises to resolve, `requestIp` method returns `undefined`

```console
➜  [arebets] curl "localhost:3000/works"     
Your IP is ::1%
➜  [arebets] curl "localhost:3000/fails"
Your IP is undefined%
```

### Workaround

For now I've simply copied the functionality from `elysia-ip` and changed the hook to `onRequest`, while also adding `ip` to the Context's `store` (you have to declare an initial value using `state` to get access to this magic):

<details>

<summary>code to make elysia-rate-limiter work correctly</summary>

```ts
// elysia_ip.ts

import type { ElysiaApp } from "./main";
...
export const ip =
    (
        config: {
            /**
             * Customize headers to check for IP address
             */
            checkHeaders?: IPHeaders[];
        } = {},
    ) =>
    (app: ElysiaApp) => {
        return app.onRequest(({ request, store }) => {
            let ip = "";
            if (globalThis.Bun) {
                if (!app.server)
                    throw new Error(
                        "Elysia server is not initialized. Make sure to call Elyisa.listen()",
                    );
                const clientIP = app.server.requestIP(request);
                ip = clientIP ? clientIP.address : "";
            }
            if (ip === "") {
                const clientIP = getIP(request.headers, config.checkHeaders);
                ip = clientIP ? clientIP : "";
            }
            store.ip = ip;
        });
    };

//main.ts

...
const app = new Elysia().state("ip", "");

app.use(ip())
    .use(
        rateLimit({
            generator: (_) => app.store.ip,
        }),
    )
    .post("/hi", () => "hello!", {
        body: t.Object({
            check: t.String()
        })
    }
    .listen(process.env.PORT || 3000);

export type ElysiaApp = typeof app;
```

</details>

**Comment by gaurishhs** [comment]
Hi, `elysia-ip` creator here, `elysia-ip` uses derive because `onRequest` does not register a new type on the Context since it relies on [PreContext](https://elysiajs.com/life-cycle/request.html#pre-context) and that's probably the reason you need to initialize it with an empty value.

**Comment by a-rebets** [comment]
Thanks, @gaurishhs ! Now I get it, those contexts are different and changing context properties inside `onRequest` won't actually pass it further into the hook chain. So the recommendation from docs is specifically for blocking - you either block and return to the client, or allow the request through.

@SaltyAom by the way, would you agree this deserves to be highlighted in the docs? smth like this for example:
<img width="557" alt="doc_change" src="https://github.com/elysiajs/elysia/assets/34752993/bb3edbf4-53ff-4d60-97d5-67170e155b4f">

About initializing with an empty value - that was solely for not losing type safety, cause you can't assign new properties to the context's `store` in a type safe way if nothing was "stored" to begin with using the `state` method. At least that's my understanding.

And with `requestIP`... well I guess one more workaround that is valid _right now_ is to store IP address in the Context using the `onParse` hook - that's for folks working with POST requests which usually have some body that will be parsed and validated. This is a hacky workaround indeed, but it's the only way to make sure no "awaits" happen before we access Bun's `requestIP` so that it does not return `undefined`. IP-related code will be processed _before_ any Elysia's built-in body parsers (see Parse phase of the lifecycle in docs)

<details>

<summary>modified code for elysia-rate-limit</summary>

```ts
// elysia_ip.ts

import type Elysia from "elysia";
...
export const ip =
    (
        config: {
            /**
             * Customize headers to check for IP address
             */
            checkHeaders?: IPHeaders[];
        } = {},
    ) =>
    (app: Elysia) => {
        return app.state("ip", "").onParse(({ request, store }) => {
            let ip = "";
            if (globalThis.Bun) {
                if (!app.server)
                    throw new Error(
                        "Elysia server is not initialized. Make sure to call Elyisa.listen()",
                    );
                const clientIP = app.server.requestIP(request);
                ip = clientIP ? clientIP.address : "";
            }
            if (ip === "") {
                const clientIP = getIP(request.headers, config.checkHeaders);
                ip = clientIP ? clientIP : "";
            }
            store.ip = ip;
        });
    };

//main.ts

...
const app = new Elysia().use(ip());

app.use(
        rateLimit({
            generator: (_) => app.store.ip,
        }),
    )
    .post("/hi", () => "hello!", {
        body: t.Object({
            check: t.String()
        })
    }
    .listen(process.env.PORT || 3000);
```

</details>

**Comment by SaltyAom** [comment]
Blocking on https://github.com/oven-sh/bun/issues/6613 with the following code:

```ts
Bun.serve({
	port: 3000,
	async fetch(request, server) {
		await Bun.sleep(10)

		console.log(server.requestIP(request)?.address)

		return new Response('a')
	}
})
```

**Comment by kravetsone** [comment]
Fixed by Bun@1.1.27

```ts
import { Elysia, t } from "elysia";
const app = new Elysia();
app
    // Always available
    .onRequest(({ request }) => {
        console.log(app.server!.requestIP(request)?.address);
    })
    // Not available when using body validator
    .onTransform(({ request }) => {
        console.log(app.server!.requestIP(request)?.address); // Not available when using body validator
    })
    // Similar to onTransform
    .derive(({ request }) => ({ ip: app.server!.requestIP(request)?.address }))
    // foo ::ffff:127.0.0.1
    .post("/foo/:id", ({ ip }) => `foo ${ip}`, {
        query: t.Object({ name: t.String() }), // Query and params do not affect the issue
        params: t.Object({ id: t.Numeric() }),
    })
    // bar undefined
    .post("/bar", ({ ip }) => `bar ${ip}`, {
        body: t.Object({ num: t.Integer() }),
    })
    .listen(3000);
```

![image](https://github.com/user-attachments/assets/d0afa749-5efd-46ed-9065-e17bdfeb8638)
